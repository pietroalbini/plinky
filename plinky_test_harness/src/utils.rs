use anyhow::{bail, Error};
use std::path::{Path, PathBuf};
use std::process::Command;

pub fn run(command: &mut Command) -> Result<(), Error> {
    let cmd_repr = format!("{command:?}");
    let output = command.output()?;
    if !output.status.success() {
        eprintln!("Failed to execute {cmd_repr}");
        eprintln!();
        eprintln!("=== stdout ===");
        eprintln!("{}", String::from_utf8_lossy(&output.stdout));
        eprintln!();
        eprintln!("=== stderr ===");
        eprintln!("{}", String::from_utf8_lossy(&output.stderr));
        eprintln!();
        bail!("command failed with exit {}", output.status);
    }
    Ok(())
}

pub struct RunAndSnapshot {
    name: String,
    path: PathBuf,
    output: String,
}

impl RunAndSnapshot {
    pub fn new(name: &str, path: &Path) -> Self {
        Self { name: name.into(), path: path.into(), output: String::new() }
    }

    pub fn run(&mut self, action: &str, command: &mut Command) -> Result<bool, Error> {
        self.separator();
        match command.output() {
            Ok(output) => {
                self.output.push_str(&format!("{action} exited with {}\n", output.status));
                for (name, content) in [("stdout", &output.stdout), ("stderr", &output.stderr)] {
                    if content.trim_ascii().is_empty() {
                        self.output.push_str(&format!("\nno {name} present\n"));
                    } else {
                        let mut content = String::from_utf8_lossy(content).to_string();
                        content = content.replace(env!("CARGO_MANIFEST_DIR"), "${project}");

                        self.output
                            .push_str(&format!("\n=== {name} ===\n{}\n", content.trim_ascii_end()));
                    }
                }
                Ok(output.status.success())
            }
            Err(err) => {
                self.output.push_str(&format!("{action} failed to execute with error: {err}"));
                Ok(false)
            }
        }
    }

    pub fn note(&mut self, note: &str) {
        self.separator();
        self.output.push_str(note);
        if !note.ends_with('\n') {
            self.output.push('\n');
        }
    }

    fn separator(&mut self) {
        if !self.output.is_empty() {
            self.output.push_str("\n==============\n\n");
        }
    }

    pub fn persist(self) {
        let mut insta_settings = insta::Settings::clone_current();
        insta_settings.set_prepend_module_to_snapshot(false);
        insta_settings.set_omit_expression(true);
        insta_settings.set_snapshot_path(self.path.canonicalize().unwrap());

        insta_settings.bind(|| {
            insta::assert_snapshot!(self.name, self.output.trim_ascii_end());
        });
    }
}

pub(crate) fn file_name(path: impl AsRef<Path>) -> String {
    path.as_ref().file_name().unwrap().to_str().unwrap().to_string()
}

#[track_caller]
pub(crate) fn err_str<T>(result: Result<T, Error>) -> Result<T, String> {
    match result {
        Ok(ok) => Ok(ok),
        Err(err) => {
            let mut repr = format!("error: {err}\n");
            let mut source = err.source();
            while let Some(err) = source {
                repr.push_str(&format!("  cause: {err}\n"));
                source = err.source();
            }
            panic!("{repr}")
        }
    }
}
