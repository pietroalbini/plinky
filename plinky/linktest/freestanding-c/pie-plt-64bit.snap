---
source: plinky_test_harness/src/utils.rs
---
linking exited with exit status: 0

no stdout present

=== stderr ===
debug print: loaded object
 │
 │  section .text#0 (perms: rx) in hello.o
 │   │
 │   │  ╭─────────────────────────────────────────────────┬──────────────────╮
 │   │  │ 55 48 89 e5 ba 0c 00 00 00 48 8d 05 00 00 00 00 │ UH.......H...... │
 │   │  │ 48 89 c6 bf 01 00 00 00 e8 00 00 00 00 bf 00 00 │ H............... │
 │   │  │ 00 00 e8 00 00 00 00                            │ .......          │
 │   │  ╰─────────────────────────────────────────────────┴──────────────────╯
 │   │
 │   │    Relocations:
 │   │  ╭────────────┬───────────────────┬────────┬────────────────────╮
 │   │  │ Type       │ Symbol            │ Offset │ Addend             │
 │   │  ├────────────┼───────────────────┼────────┼────────────────────┤
 │   │  │ Relative32 │ <section .rodata> │ 0xc    │ 0xfffffffffffffffc │
 │   │  │ PLT32      │ write             │ 0x19   │ 0xfffffffffffffffc │
 │   │  │ PLT32      │ exit              │ 0x23   │ 0xfffffffffffffffc │
 │   │  ╰────────────┴───────────────────┴────────┴────────────────────╯
 │   ┴
 │
 │  section .text#1 (perms: rx) in syscall.write.o
 │   │
 │   │  ╭─────────────────────────────────────────────────┬──────────────────╮
 │   │  │ 50 51 41 53 48 c7 c0 01 00 00 00 0f 05 41 5b 59 │ PQASH........A[Y │
 │   │  │ 58 c3                                           │ X.               │
 │   │  ╰─────────────────────────────────────────────────┴──────────────────╯
 │   ┴
 │
 │  section .text#2 (perms: rx) in syscall.exit.o
 │   │
 │   │  ╭────────────────────────────┬───────────╮
 │   │  │ 48 c7 c0 3c 00 00 00 0f 05 │ H..<..... │
 │   │  ╰────────────────────────────┴───────────╯
 │   ┴
 ┴

debug print: relocations analysis
 │
 │  global offset table .got.plt
 │   │
 │   │    Symbols:
 │   │  ╭───────┬─────────────╮
 │   │  │ Name  │ Resolved at │
 │   │  ├───────┼─────────────┤
 │   │  │ write │ runtime     │
 │   │  │ exit  │ runtime     │
 │   │  ╰───────┴─────────────╯
 │   ┴
 ┴

debug print: calculated layout
 │
 │    Layout:
 │  ╭────────────────────┬─────────────┬─────────────┬────────────────┬───────────────╮
 │  │ Part               │ File offset │ File length │ Memory address │ Memory length │
 │  ├────────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ <elf header>       │ 0x0         │ 0x40        │ 0x0            │ 0x40          │
 │  │ <program header>   │ 0x40        │ 0x230       │ 0x40           │ 0x230         │
 │  │ .interp            │ 0x270       │ 0x53        │ 0x270          │ 0x53          │
 │  │ .dynstr            │ 0x2c3       │ 0xd         │ 0x2c3          │ 0xd           │
 │  │ .dynsym            │ 0x2d0       │ 0x48        │ 0x2d0          │ 0x48          │
 │  │ .hash              │ 0x318       │ 0x18        │ 0x318          │ 0x18          │
 │  │ .gnu.hash          │ 0x330       │ 0x24        │ 0x330          │ 0x24          │
 │  │ .dynamic           │ 0x354       │ 0xc0        │ 0x354          │ 0xc0          │
 │  │ .rela.plt          │ 0x414       │ 0x30        │ 0x414          │ 0x30          │
 │  │ <section header>   │ 0x444       │ 0x500       │ -              │ -             │
 │  │ .shstrtab          │ 0x944       │ 0xa4        │ -              │ -             │
 │  │ .strtab            │ 0x9e8       │ 0x33        │ -              │ -             │
 │  │ .symtab            │ 0xa1b       │ 0x90        │ -              │ -             │
 │  │ .comment#0         │ 0xaab       │ 0x22        │ -              │ -             │
 │  │ <padding>          │ 0xacd       │ 0x533       │ -              │ -             │
 │  │ .rodata            │ 0x1000      │ 0xd         │ 0x1000         │ 0xd           │
 │  │ .eh_frame          │ 0x100d      │ 0x38        │ 0x100d         │ 0x38          │
 │  │ <padding>          │ 0x1045      │ 0xfbb       │ -              │ -             │
 │  │ .text#0            │ 0x2000      │ 0x27        │ 0x2000         │ 0x27          │
 │  │ .text#1            │ 0x2027      │ 0x12        │ 0x2027         │ 0x12          │
 │  │ .text#2            │ 0x2039      │ 0x9         │ 0x2039         │ 0x9           │
 │  │ .plt               │ 0x2042      │ 0x30        │ 0x2042         │ 0x30          │
 │  │ <padding>          │ 0x2072      │ 0xf8e       │ -              │ -             │
 │  │ .got.plt           │ 0x3000      │ 0x28        │ 0x3000         │ 0x28          │
 │  │ .note.gnu.property │ 0x3028      │ 0x30        │ 0x3028         │ 0x30          │
 │  ╰────────────────────┴─────────────┴─────────────┴────────────────┴───────────────╯
 │
 │    Segments:
 │  ╭────────┬────────┬────────────────┬───────┬────────────────────╮
 │  │ Start  │ Align  │ Type           │ Perms │ Content            │
 │  ├────────┼────────┼────────────────┼───────┼────────────────────┤
 │  │ 0x40   │ 0x1000 │ program header │ R     │ <program header>   │
 │  │ 0x270  │ 0x1    │ interpreter    │ R     │ .interp            │
 │  │ 0x0    │ 0x1000 │ program        │ R     │ <elf header>       │
 │  │        │        │                │       │ <program header>   │
 │  │        │        │                │       │ .interp            │
 │  │        │        │                │       │ .dynstr            │
 │  │        │        │                │       │ .dynsym            │
 │  │        │        │                │       │ .hash              │
 │  │        │        │                │       │ .gnu.hash          │
 │  │        │        │                │       │ .dynamic           │
 │  │        │        │                │       │ .rela.plt          │
 │  │ 0x1000 │ 0x1000 │ program        │ R     │ .rodata            │
 │  │        │        │                │       │ .eh_frame          │
 │  │ 0x2000 │ 0x1000 │ program        │ RX    │ .text#0            │
 │  │        │        │                │       │ .text#1            │
 │  │        │        │                │       │ .text#2            │
 │  │        │        │                │       │ .plt               │
 │  │ 0x3000 │ 0x1000 │ program        │ RW    │ .got.plt           │
 │  │ 0x354  │ 0x8    │ dynamic        │ R     │ .dynamic           │
 │  │ -      │ 0x1    │ GNU stack      │ RW    │                    │
 │  │ 0x3028 │ 0x8    │ GNU property   │ R     │ .note.gnu.property │
 │  │ 0x3028 │ 0x8    │ notes          │ R     │ .note.gnu.property │
 │  ╰────────┴────────┴────────────────┴───────┴────────────────────╯
 ┴

debug print: object after relocations are applied
 │
 │  section .got.plt (perms: rw) in <plinky>
 │   │
 │   │  address: 0x3000
 │   │
 │   │  ╭─────────────────────────────────────────────────┬──────────────────╮
 │   │  │ 54 03 00 00 00 00 00 00 00 00 00 00 00 00 00 00 │ T............... │
 │   │  │ 00 00 00 00 00 00 00 00 58 20 00 00 00 00 00 00 │ ........X ...... │
 │   │  │ 68 20 00 00 00 00 00 00                         │ h ......         │
 │   │  ╰─────────────────────────────────────────────────┴──────────────────╯
 │   ┴
 │
 │  section .interp (perms: r) in <plinky>
 │   │
 │   │  address: 0x270
 │   │
 │   │  ╭─────────────────────────────────────────────────┬──────────────────╮
 │   │  │ 2f 6e 69 78 2f 73 74 6f 72 65 2f 6b 70 79 32 63 │ /nix/store/kpy2c │
 │   │  │ 79 64 30 35 76 64 72 36 6a 31 68 32 30 30 61 76 │ yd05vdr6j1h200av │
 │   │  │ 38 31 66 6e 6c 78 6c 31 6a 77 30 2d 67 6c 69 62 │ 81fnlxl1jw0-glib │
 │   │  │ 63 2d 32 2e 33 39 2d 35 32 2f 6c 69 62 2f 6c 64 │ c-2.39-52/lib/ld │
 │   │  │ 2d 6c 69 6e 75 78 2d 78 38 36 2d 36 34 2e 73 6f │ -linux-x86-64.so │
 │   │  │ 2e 32 00                                        │ .2.              │
 │   │  ╰─────────────────────────────────────────────────┴──────────────────╯
 │   ┴
 │
 │  section .plt (perms: rx) in <plinky>
 │   │
 │   │  address: 0x2042
 │   │
 │   │  ╭─────────────────────────────────────────────────┬──────────────────╮
 │   │  │ ff 35 c0 0f 00 00 ff 25 c2 0f 00 00 90 90 90 90 │ .5.....%........ │
 │   │  │ ff 25 c0 0f 00 00 68 00 00 00 00 e9 e0 ff ff ff │ .%....h......... │
 │   │  │ ff 25 b8 0f 00 00 68 01 00 00 00 e9 d0 ff ff ff │ .%....h......... │
 │   │  ╰─────────────────────────────────────────────────┴──────────────────╯
 │   ┴
 │
 │  section .text#0 (perms: rx) in hello.o
 │   │
 │   │  address: 0x2000
 │   │
 │   │  ╭─────────────────────────────────────────────────┬──────────────────╮
 │   │  │ 55 48 89 e5 ba 0c 00 00 00 48 8d 05 f0 ef ff ff │ UH.......H...... │
 │   │  │ 48 89 c6 bf 01 00 00 00 e8 35 00 00 00 bf 00 00 │ H........5...... │
 │   │  │ 00 00 e8 3b 00 00 00                            │ ...;...          │
 │   │  ╰─────────────────────────────────────────────────┴──────────────────╯
 │   ┴
 │
 │  section .text#1 (perms: rx) in syscall.write.o
 │   │
 │   │  address: 0x2027
 │   │
 │   │  ╭─────────────────────────────────────────────────┬──────────────────╮
 │   │  │ 50 51 41 53 48 c7 c0 01 00 00 00 0f 05 41 5b 59 │ PQASH........A[Y │
 │   │  │ 58 c3                                           │ X.               │
 │   │  ╰─────────────────────────────────────────────────┴──────────────────╯
 │   ┴
 │
 │  section .text#2 (perms: rx) in syscall.exit.o
 │   │
 │   │  address: 0x2039
 │   │
 │   │  ╭────────────────────────────┬───────────╮
 │   │  │ 48 c7 c0 3c 00 00 00 0f 05 │ H..<..... │
 │   │  ╰────────────────────────────┴───────────╯
 │   ┴
 │
 │    Symbols:
 │  ╭───────────────────────┬──────────┬─────────────────┬────────────┬────────────────╮
 │  │ Name                  │ Type     │ Source          │ Visibility │ Value          │
 │  ├───────────────────────┼──────────┼─────────────────┼────────────┼────────────────┤
 │  │ <empty>               │ none     │ <plinky>        │ local      │ <null>         │
 │  │ <section .text#0>     │ section  │ hello.o         │ local      │ .text#0        │
 │  │ <section .rodata>     │ section  │ hello.o         │ local      │ .rodata        │
 │  │ <section .plt>        │ section  │ <plinky>        │ local      │ .plt           │
 │  │ _DYNAMIC              │ none     │ <plinky>        │ local      │ .dynamic + 0x0 │
 │  │ _GLOBAL_OFFSET_TABLE_ │ none     │ <plinky>        │ local      │ .got.plt + 0x0 │
 │  │ _start                │ function │ hello.o         │ global     │ .text#0 + 0x0  │
 │  │ exit                  │ none     │ syscall.exit.o  │ global     │ .text#2 + 0x0  │
 │  │ write                 │ none     │ syscall.write.o │ global     │ .text#1 + 0x0  │
 │  ╰───────────────────────┴──────────┴─────────────────┴────────────┴────────────────╯
 │
 │    Dynamic symbols:
 │  ╭─────────┬──────┬─────────────────┬────────────┬───────────────╮
 │  │ Name    │ Type │ Source          │ Visibility │ Value         │
 │  ├─────────┼──────┼─────────────────┼────────────┼───────────────┤
 │  │ <empty> │ none │ <plinky>        │ local      │ <null>        │
 │  │ exit    │ none │ syscall.exit.o  │ global     │ .text#2 + 0x0 │
 │  │ write   │ none │ syscall.write.o │ global     │ .text#1 + 0x0 │
 │  ╰─────────┴──────┴─────────────────┴────────────┴───────────────╯
 ┴

debug print: built elf
 │
 │    Metadata:
 │  ╭────────────┬───────────────╮
 │  │ Class      │ ELF 64bit     │
 │  │ Endian     │ Little        │
 │  │ ABI        │ System V      │
 │  │ Machine    │ x86-64        │
 │  │ Type       │ Shared object │
 │  │ Entrypoint │ 0x2000        │
 │  ╰────────────┴───────────────╯
 │
 │  section .interp (address: 0x270)
 │   │
 │   │  program data | permissions: R
 │   │
 │   │  ╭─────────────────────────────────────────────────┬──────────────────╮
 │   │  │ 2f 6e 69 78 2f 73 74 6f 72 65 2f 6b 70 79 32 63 │ /nix/store/kpy2c │
 │   │  │ 79 64 30 35 76 64 72 36 6a 31 68 32 30 30 61 76 │ yd05vdr6j1h200av │
 │   │  │ 38 31 66 6e 6c 78 6c 31 6a 77 30 2d 67 6c 69 62 │ 81fnlxl1jw0-glib │
 │   │  │ 63 2d 32 2e 33 39 2d 35 32 2f 6c 69 62 2f 6c 64 │ c-2.39-52/lib/ld │
 │   │  │ 2d 6c 69 6e 75 78 2d 78 38 36 2d 36 34 2e 73 6f │ -linux-x86-64.so │
 │   │  │ 2e 32 00                                        │ .2.              │
 │   │  ╰─────────────────────────────────────────────────┴──────────────────╯
 │   ┴
 │
 │  section .dynstr (address: 0x2c3)
 │   │
 │   │    Strings table:
 │   │  ╭─────┬───────╮
 │   │  │ 0x0 │       │
 │   │  │ 0x1 │       │
 │   │  │ 0x2 │ write │
 │   │  │ 0x8 │ exit  │
 │   │  ╰─────┴───────╯
 │   ┴
 │
 │  section .dynsym (address: 0x2d0)
 │   │
 │   │    Dynamic symbol table:
 │   │  ╭───────────┬─────────┬──────┬────────────┬────────────┬────────┬──────╮
 │   │  │ Name      │ Binding │ Type │ Visibility │ Definition │ Value  │ Size │
 │   │  ├───────────┼─────────┼──────┼────────────┼────────────┼────────┼──────┤
 │   │  │ <empty>#1 │ Local   │ -    │ Default    │ Undefined  │ 0x0    │ 0x0  │
 │   │  │ write#1   │ Global  │ -    │ Default    │ .text#1    │ 0x2027 │ 0x0  │
 │   │  │ exit#1    │ Global  │ -    │ Default    │ .text#2    │ 0x2039 │ 0x0  │
 │   │  ╰───────────┴─────────┴──────┴────────────┴────────────┴────────┴──────╯
 │   ┴
 │
 │  section .hash (address: 0x318)
 │   │
 │   │  Hash table for .dynsym
 │   │
 │   │    Content:
 │   │  ╭───────────┬───────────────────╮
 │   │  │ Bucket ID │ Symbols in bucket │
 │   │  ├───────────┼───────────────────┤
 │   │  │ 0         │ exit#1            │
 │   │  │           │ write#1           │
 │   │  ╰───────────┴───────────────────╯
 │   ┴
 │
 │  section .gnu.hash (address: 0x330)
 │   │
 │   │  GNU hash table for .dynsym
 │   │  Ignored symbols: 1
 │   │
 │   │    Bloom filter (shift of 10):
 │   │  ╭──────────────────────────────────────────────────────────────────╮
 │   │  │ 1000000000000000001000000000000010000000000000010000000000000000 │
 │   │  ╰──────────────────────────────────────────────────────────────────╯
 │   │
 │   │    Content:
 │   │  ╭───────────┬───────────────────┬──────────────────╮
 │   │  │ Bucket ID │ Symbols in bucket │ Truncated hashes │
 │   │  ├───────────┼───────────────────┼──────────────────┤
 │   │  │ 0         │ write#1           │ 10a8b550         │
 │   │  │           │ exit#1            │ 7c967e3e         │
 │   │  ╰───────────┴───────────────────┴──────────────────╯
 │   ┴
 │
 │  section .dynamic (address: 0x354)
 │   │
 │   │  dynamic information | string table: .dynstr
 │   │
 │   │  ╭─────────────────────────┬────────────────╮
 │   │  │ Kind                    │ Value          │
 │   │  ├─────────────────────────┼────────────────┤
 │   │  │ String table            │ address 0x2c3  │
 │   │  │ String table size       │ 13 bytes       │
 │   │  │ Symbol table            │ address 0x2d0  │
 │   │  │ Symbol table entry size │ 24 bytes       │
 │   │  │ Hash table              │ address 0x318  │
 │   │  │ GNU hash table          │ address 0x330  │
 │   │  │ Jump PLT relocations    │ address 0x414  │
 │   │  │ PLT relocations size    │ 48 bytes       │
 │   │  │ PLT relocations type    │ RelocationsA   │
 │   │  │ PLT GOT                 │ address 0x3000 │
 │   │  │ Flags1                  │ pie            │
 │   │  │ Null                    │ -              │
 │   │  ╰─────────────────────────┴────────────────╯
 │   ┴
 │
 │  section .rela.plt (address: 0x414)
 │   │
 │   │  symbol table:       .dynsym
 │   │  applies to section: .got.plt
 │   │
 │   │    Relocations:
 │   │  ╭─────────────────┬─────────┬────────┬────────╮
 │   │  │ Type            │ Symbol  │ Offset │ Addend │
 │   │  ├─────────────────┼─────────┼────────┼────────┤
 │   │  │ X86_64_JumpSlot │ write#1 │ 0x3018 │ 0x0    │
 │   │  │ X86_64_JumpSlot │ exit#1  │ 0x3020 │ 0x0    │
 │   │  ╰─────────────────┴─────────┴────────┴────────╯
 │   ┴
 │
 │    Segments:
 │  ╭──────────────────────┬───────┬──────────┬─────────────┬──────────┬────────────────┬────────────╮
 │  │ Type                 │ Perms │ Aligment │ File offset │ File len │ Memory address │ Memory len │
 │  ├──────────────────────┼───────┼──────────┼─────────────┼──────────┼────────────────┼────────────┤
 │  │ Program header table │ R     │ 0x1000   │ 0x40        │ 0x230    │ 0x40           │ 0x230      │
 │  │ Interpreter          │ R     │ 0x1      │ 0x270       │ 0x53     │ 0x270          │ 0x53       │
 │  │ Load                 │ R     │ 0x1000   │ 0x0         │ 0x444    │ 0x0            │ 0x444      │
 │  │ Load                 │ R     │ 0x1000   │ 0x1000      │ 0x45     │ 0x1000         │ 0x45       │
 │  │ Load                 │ R X   │ 0x1000   │ 0x2000      │ 0x72     │ 0x2000         │ 0x72       │
 │  │ Load                 │ RW    │ 0x1000   │ 0x3000      │ 0x28     │ 0x3000         │ 0x28       │
 │  │ Dynamic              │ R     │ 0x8      │ 0x354       │ 0xc0     │ 0x354          │ 0xc0       │
 │  │ Note                 │ R     │ 0x8      │ 0x3028      │ 0x30     │ 0x3028         │ 0x30       │
 │  │ GNU stack            │ RW    │ 0x1      │ 0x0         │ 0x0      │ 0x0            │ 0x0        │
 │  │ GNU property         │ R     │ 0x8      │ 0x3028      │ 0x30     │ 0x3028         │ 0x30       │
 │  ╰──────────────────────┴───────┴──────────┴─────────────┴──────────┴────────────────┴────────────╯
 ┴

==============

running exited with exit status: 0

=== stdout ===
Hello world

no stderr present
